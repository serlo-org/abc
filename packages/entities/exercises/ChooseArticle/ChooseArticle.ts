// TODO: don't depend on React Native app
import { Article } from '../../../../src/assets/words';
import { Optional } from '../../../../src/types/index';
// TODO: don't depend on React Native app
import Word from '../../../../src/word';
import AbstractExercise from '../AbstractExercise';

export interface IProps {
  word: Word;
}

export type IState = Optional<Article>;

class ChooseArticle extends AbstractExercise<IProps, IState> {
  public getInitialState() {
    return undefined;
  }

  public isCorrect(state: IState) {
    const { word } = this.getProps();

    return word.getArticle() === state;
  }
}

export default ChooseArticle;
