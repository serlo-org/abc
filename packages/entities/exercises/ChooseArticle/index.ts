import Exercise from './ChooseArticle';
import fixtures from './ChooseArticle.fixtures';

export { Exercise, fixtures };
