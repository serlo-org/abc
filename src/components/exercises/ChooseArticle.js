import { addIndex, map } from 'ramda';
import React from 'react';
import { View, Text, TouchableOpacity } from 'react-native';

import speakerImage from '../../assets/images/speaker.png';
import {
  BLACK_TRANSPARENT,
  PRIMARY_STRONG,
  PRIMARY,
  MASCULINE_STRONG,
  MASCULINE_WEAK,
  FEMININE_STRONG,
  FEMININE_WEAK,
  NEUTER_STRONG,
  NEUTER_WEAK
} from '../../styles/colors';
import { DEFAULT } from '../../styles/text';
import { loadSounds, playAll } from '../helpers/audio';
import { RoundImageWithButton } from '../Components';

const mapIndexed = addIndex(map);

const articles = ['der', 'die', 'das'];

const a = {
  [articles[0]]: {
    weak: MASCULINE_WEAK,
    strong: MASCULINE_STRONG
  },
  [articles[1]]: {
    weak: FEMININE_WEAK,
    strong: FEMININE_STRONG
  },
  [articles[2]]: {
    weak: NEUTER_WEAK,
    strong: NEUTER_STRONG
  }
};

const styles = {
  letterStyle: {
    backgroundColor: PRIMARY_STRONG
  },
  article: {
    marginHorizontal: 10,
    padding: 10
  },
  shared: {
    padding: 5,
    margin: 2,
    borderRadius: 20,
    elevation: 10,
    shadowColor: BLACK_TRANSPARENT,
    shadowOpacity: 1,
    shadowRadius: 0,
    shadowOffset: {
      height: 4,
      width: 4
    }
  }
};

class ChooseArticle extends React.Component {
  constructor() {
    super();

    this.state = {
      selectedArticle: null
    };
  }

  play = () => playAll(this.props.sounds);

  selectArticle = article => () => {
    this.setState(({ selectedArticle }) => ({
      selectedArticle: selectedArticle === article ? null : article
    }));
  };

  render() {
    const letters = mapIndexed(
      (char, key) =>
        <View
          key={key}
          style={[{ padding: 5 }, styles.letterStyle, styles.shared]}
        >
          <Text style={DEFAULT}>
            {char}
          </Text>
        </View>,
      this.props.text
    );

    const articleButtons = mapIndexed(
      (article, key) =>
        <TouchableOpacity
          key={key}
          style={[
            styles.shared,
            styles.article,
            {
              backgroundColor:
                this.state.selectedArticle === article
                  ? a[article].strong
                  : a[article].weak
            }
          ]}
          onPress={this.selectArticle(article)}
        >
          <Text style={DEFAULT}>
            {article}
          </Text>
        </TouchableOpacity>,
      articles
    );

    return (
      <View
        style={{
          flex: 1,
          backgroundColor: PRIMARY,
          alignItems: 'center',
          justifyContent: 'space-around'
        }}
      >
        <RoundImageWithButton
          image={this.props.image}
          imageSize={200}
          icon={speakerImage}
          buttonSize={40}
          onPress={this.play}
        />
        <View style={{ flexDirection: 'row' }}>
          {letters}
        </View>
        <View style={{ flexDirection: 'row', justifyContent: 'space-between' }}>
          {articleButtons}
        </View>
      </View>
    );
  }
}

export default loadSounds(ChooseArticle);
