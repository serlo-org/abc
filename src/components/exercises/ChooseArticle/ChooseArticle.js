import { addIndex, map } from 'ramda';
import React, { Component } from 'react';
import { View, Text, TouchableOpacity } from 'react-native';

import { getWord } from '../../../helpers/words';
import {
  BLACK_TRANSPARENT,
  PRIMARY_STRONG,
  PRIMARY,
  MASCULINE_STRONG,
  MASCULINE_WEAK,
  FEMININE_STRONG,
  FEMININE_WEAK,
  NEUTER_STRONG,
  NEUTER_WEAK
} from '../../../styles/colors';
import { DEFAULT } from '../../../styles/text';
import WordImageWithSounds from '../../common/WordImageWithSounds';

const mapIndexed = addIndex(map);

const articles = ['der', 'die', 'das'];

const a = {
  [articles[0]]: {
    weak: MASCULINE_WEAK,
    strong: MASCULINE_STRONG
  },
  [articles[1]]: {
    weak: FEMININE_WEAK,
    strong: FEMININE_STRONG
  },
  [articles[2]]: {
    weak: NEUTER_WEAK,
    strong: NEUTER_STRONG
  }
};

const styles = {
  letterStyle: {
    backgroundColor: PRIMARY_STRONG
  },
  article: {
    marginHorizontal: 10,
    padding: 10
  },
  shared: {
    padding: 5,
    margin: 2,
    borderRadius: 20,
    elevation: 10,
    shadowColor: BLACK_TRANSPARENT,
    shadowOpacity: 1,
    shadowRadius: 0,
    shadowOffset: {
      height: 4,
      width: 4
    }
  }
};

class ChooseArticle extends Component {
  selectArticle = article => () => {
    this.props.setState(state => (state === article ? '' : article));
  };

  render() {
    const letters = mapIndexed(
      (char, key) => (
        <View
          key={key}
          style={[{ padding: 5 }, styles.letterStyle, styles.shared]}
        >
          <Text style={DEFAULT}>{char}</Text>
        </View>
      ),
      getWord(this.props.word)
    );

    const articleButtons = mapIndexed(
      (article, key) => (
        <TouchableOpacity
          key={key}
          style={[
            styles.shared,
            styles.article,
            {
              backgroundColor:
                this.props.state === article
                  ? a[article].strong
                  : a[article].weak
            }
          ]}
          onPress={this.selectArticle(article)}
        >
          <Text style={DEFAULT}>{article}</Text>
        </TouchableOpacity>
      ),
      articles
    );

    return (
      <View
        style={{
          flex: 1,
          alignItems: 'center',
          justifyContent: 'space-around'
        }}
      >
        <WordImageWithSounds word={this.props.word} longSound />
        <View style={{ flexDirection: 'row' }}>{letters}</View>
        <View style={{ flexDirection: 'row', justifyContent: 'space-between' }}>
          {articleButtons}
        </View>
      </View>
    );
  }
}

export default ChooseArticle;
