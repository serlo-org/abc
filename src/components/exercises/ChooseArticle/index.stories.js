import { storiesOf } from '@storybook/react-native';

import * as ChooseArticle from '.';
import { createStoriesFromFixtures } from '../helpers';

const story = storiesOf('exercises/ChooseArticle', module);

createStoriesFromFixtures(story, ChooseArticle);
