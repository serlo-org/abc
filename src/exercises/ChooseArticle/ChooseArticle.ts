import { getArticle } from '../../helpers/words.js';
import AbstractExercise from '../AbstractExercise';

export interface IProps {
  word: any;
}

export type IState = string;

class ChooseArticle extends AbstractExercise<IProps, IState> {
  public getInitialState() {
    return '' as IState;
  }

  public isCorrect(state: IState) {
    const { word } = this.getProps();
    const articleString = getArticle(word).toUpperCase();

    return articleString === state.toUpperCase();
  }
}

export default ChooseArticle;
