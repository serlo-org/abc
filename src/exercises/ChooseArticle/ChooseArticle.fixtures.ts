import { getWordObject } from '../../helpers/words.js';
import { Fixtures } from '../helpers';
import { IProps, IState } from './ChooseArticle';

const fixtures: Fixtures<IProps, IState> = [
  {
    name: 'correct article ananas',
    props: {
      word: getWordObject('ananas')
    },
    state: 'die',
    isCorrect: true
  },
  {
    name: 'wrong article apfel',
    props: {
      word: getWordObject('apfel')
    },
    state: 'die',
    isCorrect: false
  }
];

export default fixtures;
