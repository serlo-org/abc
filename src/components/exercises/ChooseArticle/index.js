import { ChooseArticle } from '../../../../packages/entities/exercises';
import Component from './ChooseArticle';

export const Exercise = ChooseArticle.Exercise;
export const fixtures = ChooseArticle.fixtures;
export { Component };
