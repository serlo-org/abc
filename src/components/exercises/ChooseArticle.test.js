import React from 'react';
import renderer from 'react-test-renderer';

import ChooseArticle from './ChooseArticle';

it('renders without crashing', () => {
  const tree = renderer.create(
    <ChooseArticle
      image={require('../../assets/images/ananas.jpg')}
      sounds={[
        require('../../assets/sounds/ananas_short.mp3'),
        require('../../assets/sounds/ananas_long.mp3')
      ]}
      text="Ananas"
    />
  );
  expect(tree).toMatchSnapshot();
});
