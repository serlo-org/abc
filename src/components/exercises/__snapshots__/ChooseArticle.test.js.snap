// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#00b4d5",
      "flex": 1,
      "justifyContent": "space-around",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.1)",
          "borderRadius": 9999999,
          "margin": 20,
        }
      }
    >
      <Image
        resizeMode="cover"
        source={1}
        style={
          Object {
            "borderRadius": 100,
            "height": 200,
            "margin": 20,
            "width": 200,
          }
        }
      />
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "marginLeft": -70,
          "marginRight": 20,
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Object {
            "backgroundColor": "#73dbff",
            "borderRadius": 9999,
            "elevation": 10,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
          }
        }
      >
        <Image
          source={1}
          style={
            Object {
              "height": 40,
              "resizeMode": "contain",
              "width": 40,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "padding": 5,
          },
          Object {
            "backgroundColor": "#17bcde",
          },
          Object {
            "borderRadius": 20,
            "elevation": 10,
            "margin": 2,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "norddruck",
            "fontSize": 35,
            "textAlign": "center",
          }
        }
      >
        A
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "padding": 5,
          },
          Object {
            "backgroundColor": "#17bcde",
          },
          Object {
            "borderRadius": 20,
            "elevation": 10,
            "margin": 2,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "norddruck",
            "fontSize": 35,
            "textAlign": "center",
          }
        }
      >
        n
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "padding": 5,
          },
          Object {
            "backgroundColor": "#17bcde",
          },
          Object {
            "borderRadius": 20,
            "elevation": 10,
            "margin": 2,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "norddruck",
            "fontSize": 35,
            "textAlign": "center",
          }
        }
      >
        a
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "padding": 5,
          },
          Object {
            "backgroundColor": "#17bcde",
          },
          Object {
            "borderRadius": 20,
            "elevation": 10,
            "margin": 2,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "norddruck",
            "fontSize": 35,
            "textAlign": "center",
          }
        }
      >
        n
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "padding": 5,
          },
          Object {
            "backgroundColor": "#17bcde",
          },
          Object {
            "borderRadius": 20,
            "elevation": 10,
            "margin": 2,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "norddruck",
            "fontSize": 35,
            "textAlign": "center",
          }
        }
      >
        a
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "padding": 5,
          },
          Object {
            "backgroundColor": "#17bcde",
          },
          Object {
            "borderRadius": 20,
            "elevation": 10,
            "margin": 2,
            "padding": 5,
            "shadowColor": "rgba(0, 0, 0, 0.1)",
            "shadowOffset": Object {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 1,
            "shadowRadius": 0,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "norddruck",
            "fontSize": 35,
            "textAlign": "center",
          }
        }
      >
        s
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#6666fe",
          "borderRadius": 20,
          "elevation": 10,
          "margin": 2,
          "marginHorizontal": 10,
          "opacity": 1,
          "padding": 10,
          "shadowColor": "rgba(0, 0, 0, 0.1)",
          "shadowOffset": Object {
            "height": 4,
            "width": 4,
          },
          "shadowOpacity": 1,
          "shadowRadius": 0,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "norddruck",
            "fontSize": 35,
            "textAlign": "center",
          }
        }
      >
        der
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#fe6666",
          "borderRadius": 20,
          "elevation": 10,
          "margin": 2,
          "marginHorizontal": 10,
          "opacity": 1,
          "padding": 10,
          "shadowColor": "rgba(0, 0, 0, 0.1)",
          "shadowOffset": Object {
            "height": 4,
            "width": 4,
          },
          "shadowOpacity": 1,
          "shadowRadius": 0,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "norddruck",
            "fontSize": 35,
            "textAlign": "center",
          }
        }
      >
        die
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      isTVSelectable={true}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#67af66",
          "borderRadius": 20,
          "elevation": 10,
          "margin": 2,
          "marginHorizontal": 10,
          "opacity": 1,
          "padding": 10,
          "shadowColor": "rgba(0, 0, 0, 0.1)",
          "shadowOffset": Object {
            "height": 4,
            "width": 4,
          },
          "shadowOpacity": 1,
          "shadowRadius": 0,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "fontFamily": "norddruck",
            "fontSize": 35,
            "textAlign": "center",
          }
        }
      >
        das
      </Text>
    </View>
  </View>
</View>
`;
